doozerlib/metadata.py

```diff
@@ -190,37 +294,109 @@ def cgit_url(self, filename: str, commit_hash: Optional[str] = None, branch: Opt
-                    return default
-                raise IOError("No builds detected for %s using tag: %s" % (self.qualified_name, self.candidate_brew_tag()))
-            return builds[0]
-
-    def get_latest_build_info(self, default=-1):
+            package_info = koji_api.getPackage(package_name)  # e.g. {'id': 66873, 'name': 'atomic-openshift-descheduler-container'}
+            if not package_info:
+                raise IOError(f'No brew package is defined for {package_name}')
+            package_id = package_info['id']  # we could just constrain package name using pattern glob, but providing package ID # should be a much more efficient DB query.
+            # listBuilds returns all builds for the package; We need to limit the query to the builds
+            # relevant for our major/minor.
+
+            # RPMs do not have a 'v' in front of their version; images do.
+            if self.meta_type == 'image':
+                ver_prefix = 'v'  # openshift-enterprise-console-container-v4.7.0-202106032231.p0.git.d9f4379
+            else:
+                ver_prefix = ''  # openshift-clients-4.7.0-202106032231.p0.git.e29b355.el8
+
+            pattern_prefix = f'{package_name}-{ver_prefix}{self.branch_major_minor()}.'
+
+            if assembly is None:
+                assembly = self.runtime.assembly
+
+            def latest_build_list(pattern_suffix):
+                # We need '*' after pattern_suffix for RPMs, which have an extra .el7/.el8 on their
+                # release field, normally.
+                builds = koji_api.listBuilds(packageID=package_id,
+                                             state=None if build_state is None else build_state.value,
+                                             pattern=f'{pattern_prefix}{extra_pattern}{pattern_suffix}*',
+                                             queryOpts={'limit': 1, 'order': '-creation_event_id'})
+
+                # Ensure the suffix ends the string OR at least terminated by a '.' .
+                # This latter check ensures that 'assembly.how' doesn't not match a build from
+                # "assembly.howdy'.
+                refined = [b for b in builds if b['nvr'].endswith(pattern_suffix) or f'{pattern_suffix}.' in b['nvr']]
+
+                if refined:
+                    # A final sanity check to see if the build is tagged with something we
+                    # respect. There is a chance that a human may untag a build. There
+                    # is no standard practice at present in which they should (they should just trigger
+                    # a rebuild). If we find the latest build is not tagged appropriately, blow up
+                    # and let a human figure out what happened.
+                    check_nvr = refined[0]['nvr']
+                    tags = koji_api.listTags(build=check_nvr, pattern=f'{self.branch()}*')
+                    if not tags:
```

Comment:
A simple fix for this error could be checking if the intersection of `tags` and `self.candidate_brew_tags()` is not empty.

```suggestion
                    tags = {tag['name'] for tag in koji_api.listTags(build=check_nvr)}
                    if not tags & set(self.candidate_brew_tags()):
```

Reply:
That put me on the right track. In practice though, if we are looking for a `.assembly.hotfix` and find a valid `assembly.stream`, it will not be tagged with `-hotfix`. I've expanded acceptable tags to be `-candidate` and `-hotfix`.